### ISSUE TITLE
More fine-grained control on callback for raftstore

### ISSUE THREAD

**Issue by BusyJay** [tikv/tikv#9113]
## Bug Report

We currently rely on raft to handle callback of requests: a callback will be invoked when the corresponding entry is applied/committed or overwritten.  However, raft doesn't guarantee when will an entry be committed or overwritten, so the time when a callback is invoked is unpredictable. If a leader loses its leadership without rejoining the group or it becomes leader but takes a long time to commit its first log, any accepted requests will stuck. For the transaction layer, pending requests will occupy hash slots and lead to ServerIsBusy eventually due to conflict collisions.

I propose to add timeout check on callbacks. If a command is timeout, it should be invoked early using a `Timeout` error. Client should retry in this situation. When raftstore decides to return `Timeout` error, it should not serve any Read requests until it has applied to the `last_index`.

As a simple and quick implementation that just solves the problem of slow election, if a peer becomes follower from candidate, it can clear all the pending commands. Change from candidate to follower means it fails to campaign. If it was leader, its uncommitted logs may be overwritten by new leader.

**Comment by youjiali1995** [comment]
@MyonKeminta @sticnarf PTAL. It adds another undetermined error to transaction commit.

**Comment by cfzjywxk** [comment]
It's reasonable to return `timeout` or `undetermined ` error. Maybe processing of different comman types for `timeout` error should be considered on the client side.

**Comment by sticnarf** [comment]
> If a leader loses its leadership without rejoining the group or it becomes leader but takes a long time to commit its first log, any accepted requests will stuck. For the transaction layer, pending requests will occupy hash slots and lead to ServerIsBusy eventually due to either real conflict or hash collisions.

What could be the cause of " lose its leadership without rejoining the group or it becomes leader but takes a long time to commit its first log"? If it's due to a persistent network error, it's unlikely that it could receive more requests until `ServerIsBusy`.

**Comment by BusyJay** [comment]
Yes. Or a replica is removed without acknowledgement, or one follower is down and another follower is waiting for snapshot.

**Comment by MyonKeminta** [comment]
I have a question. Currently we now release latch in scheduler by letting raftstore call a callback on apply finish. Consider this case:

1. Request A tries to write k1, acquired the latch, and waiting for apply finish
2. Request A timed out.
3. Request B acquires tries to write k1 again

What should we do in this case?

2.a Request A responds timeout error to the client, but still waiting apply finishing. This seems didn't solve the problem in this issue.

2.b Request A abandoned the callback and released the latch.
3.b Request B successfully acquires the latch and got a snapshot (in which request A haven't apply). As an example, request A can be a prewrite operation, and request B is a prewrite operation of another transaction.
4.b Request B doesn't see request A's result so it writes down a Lock
5.b Later, request A and request B's writes are applied successively. A's Lock is overwritten. Then the transaction that request A belongs to would be broken.

I think both choice is not acceptable.

**Comment by MyonKeminta** [comment]
> When raftstore decides to return Timeout error, it should not serve any Read requests until it has applied to the last_index.

Does this means the region can't provide snapshots in this case? If so, this seems also makes other keys that's not being written unreadable. Is this ok?

**Comment by BusyJay** [comment]
> Does this means the region can't provide snapshots in this case?

Yes, it is.

> Is this ok?

I'm afraid it's the trade off between write unavailability for the whole node or read/write unavailability for just one region.

A alternative solution to the problem is keep current raft implementation, but split latch into region based. In this case, only write is unavailable for single region as expected and perhaps easier to be implemented correctly. The price is memory.

**Comment by youjiali1995** [comment]
There is a problem during rolling update. Old version TiDBs will think it fails to commit a transaction when encounters this error. It will clean up all locks but maybe the transaction is committed. @BusyJay

**Comment by sticnarf** [comment]
> There is a problem during rolling update. Old version TiDBs will think it fails to commit a transaction when encounters this error. It will clean up all locks but maybe the transaction is committed. @BusyJay

We may let TiKV return an RPC error, so we can follow the original undetermined logic.

**Comment by BusyJay** [comment]
If we really care about undetermined errors, we should use a standard flag to show if it's undetermined. And we still need to polish tikv client to distinguish if an error is deterministic. I'm afraid we don't have any official statements about whether a transaction *really* fails, deterministic error should be a new feature instead of something this issue can break.

**Comment by youjiali1995** [comment]
Yes. No need to handle compatible issues.

**Comment by BusyJay** [comment]
We have verify the new latch algorithm is very unlikely to cause conflict, so I  lower the severity of the issue.

**Comment by cfzjywxk** [comment]
@youjiali1995 @BusyJay 
Maybe we should change the severity to `major` as some customer have encountered such issues in which raftstore get stuck and the `pending write bytes` is always at a high value exceeding the rate limit value.

**Comment by BusyJay** [comment]
What's the TiKV version?

**Comment by cfzjywxk** [comment]
> What's the TiKV version?

v5.0-rc, the deadlock [issue](https://github.com/tikv/tikv/issues/9044) is encoutered. RaftStore threads get stuck because of the deadlock or the lost of the majority of repicas on the stucked tikv nodes, these tikv nodes could not continue to work.

**Comment by BusyJay** [comment]
I'm afraid raftstore can do nothing when it's block by deadlock. Perhaps additional protections need to be added to scheduler layer.
